### DISCUSSION TITLE
Anyway to import dice stats?

### DISCUSSION THREAD

**Discussion by gralamin** [jacobwojoski/dice-stats#150] [Q&A] [answered]
I have a whole bunch of manually tracked dice stats (Just d20 rolls) from before I moved to foundry. Any easy way to import some of this data? It isn't one to one with the format, but I have:

```
player, roll value, skill name / attack / flat / other, session number
Berbuil,17,Athletics,63
```

I'm a programmer, so no issues if I need to do some coding. My guess is that I need to:
1. Create a "roll info parser" for my data to get the DS_MSG_ROLL_INFO object formats
2. Get the dice stats instance
3. Map each player to your player
4. repeat the loop on dice-stats-main-js line 270?
5. Manually call the save player data?

Is that roughly correct?

**Comment by jacobwojoski** [comment]
That sounds like it should work well. You can make a macro that calls the api functions. I’ll take a better look tomorrow when I have some time and give a more detailed response . 

Make sure to look at the api file and it should give some examples on how to use it. You can probably just call add_roll(data)

**Comment by jacobwojoski** [comment]
if the system your using doesn't have sub categories you can save all the rolls using `saveRollValue({STRING:Player_id}, {ENUM:DIE_TYPE: #}, {INT: #}) -> Save a roll result to a player`  
If the system tracks saves, attacks, etc you'll need to build both classes from dice-stats-rollmsginfo.js and fill in the object manually for every roll before calling addRoll(PlayerID, MS_MSG_DATA)

Imo the mod is designed and best used for per session stats as roll results all lead to averages anyway. I find the long term stats not very interseting and the per session stats much more swingy and fun.

**Reply by gralamin** [reply]
That is fair, but I'm a good few thousand rolls in and we have some statistical outliers in our group, so its fun to see long term if it will actually even out or not. I of course assume it will over a long enough period, but its unclear if the period is long enough.

**Reply by gralamin** [reply]
The following seems to work on my local test instance for PF2e. In this case I mapped to the same user, won't be true with my production data, and I will expand the `forMap`, but just double checking this looks right to @jacobwojoski 

```javascript
// Convert your data to json to work with, sample object here:
// Session data can't be saved, its here for my debugging
let testData = [
    {'character': "Berbuil", 'roll': 4, 'for': 'Perception', 'session': 1},
    {'character': 'Ducas', 'roll': 9, 'for': 'Attack', 'session': 2}
  ]
  
  // Map from character name to player id
  // Easiest way to get this is to look at game.users and populate this yourself
  let userMap = {'Berbuil': '8JKMwMSn3HTpGUAV',
  'Ducas': '8JKMwMSn3HTpGUAV'}
  
  // D20 = 7. https://github.com/jacobwojoski/dice-stats/blob/main/scripts/dice-stats-globals.js#L82
  let dieType = 7;
  let dieInfoTemplate = {DieType: dieType, RollType: null, RollValue: null, IsBlind: false}
  
  // https://github.com/jacobwojoski/dice-stats/blob/main/scripts/dice-stats-globals.js#L108 and
  // https://github.com/jacobwojoski/dice-stats/blob/main/scripts/systemMessageParsers/pf2eSystemMessageParser.js#L125C1-L126C1
  // Fill this with every check type you track.
  let forMap = {
    "Attack": 0,
    "Will": 2,
    "Fortitude": 2,
    "Reflex": 2,
    "Flat": 5,
    "Athletics": 3,
    "Strength": 4,
    "Perception": 3, 
  }
  
  // Convert to the PF2E roll, we need to match a DS_MSG_ROLL_INFO from https://github.com/jacobwojoski/dice-stats/blob/main/scripts/appdatastorage/dice-stats-rollmsginfo.js
  // Assumptions:  20 are always crit success, 1 are crit failure 11+ is success, 9- is failure, 10 is unknown. This will be slightly wrong depending on how hard DCs were, but shouldn't be massively off.
  testData.forEach(data => {
    let degSuccess = 0;
    if (data['roll'] == 20) {
      // assume crit success
      degSuccess = 4;
    }
    else if (data['roll'] == 1) {
      // assume crit fail
      degSuccess = 1;
    }
    else if (data['roll'] >= 11) {
      // assume success
      degSuccess = 3;
    }
    else if (data['roll'] <= 9) {
      // assume failure
      degSuccess = 2;
    }
    // Will be unknown if its a 10 right now. Make DC 11.
    let rollType = forMap[data['for']];
    let dieInfo = {...dieInfoTemplate, 'RollType': rollType, 'RollValue': data['roll']};
    let rollInfo = {
      DiceInfo: [dieInfo],
      RollType: forMap[data['for']],
      DegSuccess: degSuccess,
      CheckDiff: data['roll'] - 11,
      UsedAdvantage: false,
      MissFromAdv: false,
      HitFromAdv: false
    }
    let player = userMap[data['character']];
    game.modules.get('dice-stats')?.api?.saveRollInfo(player, rollInfo);
  });
  ```

**Reply by gralamin** [reply]
Turns out this does NOT work, it only saves it in my local data, and doesn't make it into the database so my players can't see it.

**Reply by jacobwojoski** [reply]
Okay, I’ll make an issue I know the problem. It only saves the data to the database for any rolls that get added for your player. So the gm values should save when you run it but none of the others do. 

The temp fix would be to have each player run the macro and it might work. 

I need to make a ‘save_local_to_db()’ api function. That you can call to update the database with whatever your local data is.

**Reply by gralamin** [reply]
Looks like running as a player does not currently work either (even if I ignore rolls that aren't that player)

**Comment by jacobwojoski** [comment]
Looks Good. Can't say I've tested the API too much besides the open char sheets stuff so let me know if you find any issues.

**Comment by jacobwojoski** [comment, marked as answer]
Okay added new API functions to allow saving of all local data into the DB. API fn's are as follows. 

![image](https://github.com/user-attachments/assets/23e0e895-aba1-4b8d-afed-65489460c591)

If you add the saveAll call at the end of the macro. It should update the DB. If The players area already logged in. They won't see the changes unless they use the "Load from DB button".

If they roll a die before loading from DB it will overwrite the DB again with their data.

**Reply by gralamin** [reply]
Giving it a try!

**Reply by gralamin** [reply]
Looks to be working now!

For anyone coming to this discussion later, here is the macro for pf2e. Run as a script macro, as Game Master user:
```
// Convert your data to json to work with, sample object here:
// Session data can't be saved, its here for my debugging
let testData = [
    {'character': "Berbuil", 'roll': 4, 'for': 'Perception', 'session': 1},
    {'character': 'Ducas', 'roll': 9, 'for': 'Attack', 'session': 2}
  ]
  
  // Map from character name to player id
  // Easiest way to get this is to look at game.users and populate this yourself
  let userMap = {'Berbuil': '8JKMwMSn3HTpGUAV',
  'Ducas': '8JKMwMSn3HTpGUAV'}
  
  // D20 = 7. https://github.com/jacobwojoski/dice-stats/blob/main/scripts/dice-stats-globals.js#L82
  let dieType = 7;
  let dieInfoTemplate = {DieType: dieType, RollType: null, RollValue: null, IsBlind: false}
  
  // https://github.com/jacobwojoski/dice-stats/blob/main/scripts/dice-stats-globals.js#L108 and
  // https://github.com/jacobwojoski/dice-stats/blob/main/scripts/systemMessageParsers/pf2eSystemMessageParser.js#L125C1-L126C1
  // Fill this with every check type you track.
  let forMap = {
    "Attack": 0,
    "Will": 2,
    "Fortitude": 2,
    "Reflex": 2,
    "Flat": 5,
    "Athletics": 3,
    "Strength": 4,
    "Perception": 3, 
  }
  
  // Convert to the PF2E roll, we need to match a DS_MSG_ROLL_INFO from https://github.com/jacobwojoski/dice-stats/blob/main/scripts/appdatastorage/dice-stats-rollmsginfo.js
  // Assumptions:  20 are always crit success, 1 are crit failure 11+ is success, 9- is failure, 10 is unknown. This will be slightly wrong depending on how hard DCs were, but shouldn't be massively off.
  testData.forEach(data => {
    let degSuccess = 0;
    if (data['roll'] == 20) {
      // assume crit success
      degSuccess = 4;
    }
    else if (data['roll'] == 1) {
      // assume crit fail
      degSuccess = 1;
    }
    else if (data['roll'] >= 11) {
      // assume success
      degSuccess = 3;
    }
    else if (data['roll'] <= 9) {
      // assume failure
      degSuccess = 2;
    }
    // Will be unknown if its a 10 right now. Make DC 11.
    let rollType = forMap[data['for']];
    let dieInfo = {...dieInfoTemplate, 'RollType': rollType, 'RollValue': data['roll']};
    let rollInfo = {
      DiceInfo: [dieInfo],
      RollType: forMap[data['for']],
      DegSuccess: degSuccess,
      CheckDiff: data['roll'] - 11,
      UsedAdvantage: false,
      MissFromAdv: false,
      HitFromAdv: false
    }
    let player = userMap[data['character']];
    game.modules.get('dice-stats')?.api?.saveRollInfo(player, rollInfo);
  });
  game.modules.get('dice-stats')?.api?.saveAllLocalDataToDB();
```
